vr-logger/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+        }
+    }
+    __drop_global_logger
+};
+
+pub fn init_global_logger(file: PathBuf, format: &str) {
+    GLOBAL_LOGGER_INIT.call_once(|| unsafe {
+        let _ = &*GLOBAL_LOGGER.get_or_insert(Arc::new(
+            Logger::new(file, format).expect("failed to init logger"),
+        ));
+    })
+}
+
+#[macro_export]
+macro_rules! log {
+    (target: $target:expr, $lvl:expr, $($arg:tt)+) => {
+        $crate::global_log(
+            format_args!($($arg)+),
+            $lvl,
+        (
+            $target,
+            module_path!(),
+            file!()
+        ),
+        line!()
+        )
+    };
+    ($lvl:expr, $($arg:tt)+) => { $crate::log!(target: module_path!(), $lvl, $($arg)+) }
+}
+
+#[macro_export]
+macro_rules! error {
+    (target: $target:expr, $($arg:tt)+) => {
+        $crate::log!( $target, $crate::LogLevel::Error, $($arg)+)
+    };
+    ($($arg:tt)+) => { $crate::log!($crate::LogLevel::Error, $($arg)+) }
+}
+
+#[macro_export]
+macro_rules! warn {
+    (target: $target:expr, $($arg:tt)+) => {
+        $crate::log!( $target, $crate::LogLevel::Warn, $($arg)+)
+    };
+    ($($arg:tt)+) => { $crate::log!($crate::LogLevel::Warn, $($arg)+) }
+}
+
+#[macro_export]
+macro_rules! info {
+    (target: $target:expr, $($arg:tt)+) => {
+        $crate::log!( $target, $crate::LogLevel::Info, $($arg)+)
+    };
+    ($($arg:tt)+) => { $crate::log!($crate::LogLevel::Info, $($arg)+) }
+}
+
+#[macro_export]
+macro_rules! debug {
+    (target: $target:expr, $($arg:tt)+) => {
+        $crate::log!( $target, $crate::LogLevel::Debug, $($arg)+)
+    };
+    ($($arg:tt)+) => { $crate::log!($crate::LogLevel::Debug, $($arg)+) }
+}
+
+#[macro_export]
+macro_rules! trace {
+    (target: $target:expr, $($arg:tt)+) => {
+        $crate::log!( $target, $crate::LogLevel::Trace, $($arg)+)
+    };
+    ($($arg:tt)+) => { $crate::log!($crate::LogLevel::Trace, $($arg)+) }
+}
+
+pub fn global_log(
+    args: Arguments,
+    lvl: LogLevel,
+    (target, module, file): (&str, &str, &str),
+    line: u32,
+) {
+    if let Some(global_log) = unsafe { &GLOBAL_LOGGER } {
+        global_log.log(args, lvl, (target, module, file), line)
+    }
+}
+
+macro_rules! cyan {
+    ($str:literal) => {
+        concat!("\x1b[0;36m", $str, "\x1b[0m")
+    };
+}
+
+macro_rules! red {
+    ($str:literal) => {
+        concat!("\x1b[0;31m", $str, "\x1b[0m")
+    };
+}
+
+macro_rules! yellow {
+    ($str:literal) => {
+        concat!("\x1b[0;33m", $str, "\x1b[0m")
+    };
+}
+
+macro_rules! green {
+    ($str:literal) => {
+        concat!("\x1b[0;32m", $str, "\x1b[0m")
+    };
+}
+
+macro_rules! purple {
+    ($str:literal) => {
+        concat!("\x1b[0;35m", $str, "\x1b[0m")
+    };
+}
+
+const SECONDS_IN_A_DAY: u64 = SECONDS_IN_AN_HOUR * 24;
+const SECONDS_IN_AN_HOUR: u64 = SECONDS_IN_A_MINUTE * 60;
+const SECONDS_IN_A_MINUTE: u64 = 60;
+const DAYS_IN_A_YEAR: u64 = 365;
+
+const UNIX_EPOCH_YEAR: u64 = 1970;
+
+pub fn generate_utc_string() -> String {
```

Comment:
Are there no inbuild methods/good libraries for this? (I want to stay off of time stuff as much as possible)

Reply:
In comparison to those libraries this is more than minimal, it is also correct

Reply:
2. The only point where this is going to be used are the log messages